### PR TITLE
feature(sct): Argus test run tracking

### PR DESCRIPTION
**Pull request by k0machi**
This PR adds argus test run tracking, intended to track changes to
test runs in real-time - Resource allocation/deallocation, status
changes, nemeses starts/successes/failures, and to track test
information such as test name, group, release name and test
configuration

[Trello link](https://trello.com/c/J7XlShSk/3956-integrate-argusdb-with-sct)

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~New configuration option are added and documented (in `sdcm/sct_config.py`)~
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [x] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by k0machi** [comment]
Created a [Jenkins Staging Job](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/alexey/job/alexey-argus-testing/) to test everything

File: sct.py [hunk 1]
```diff
@@ -981,6 +985,21 @@ def collect_logs(test_id=None, logdir=None, backend=None, config_file=None):
 
     click.echo(table.get_string(title="Collected logs by test-id: {}".format(collector.test_id)))
 
+    try:
+        test_run = ArgusTestRun.get(test_id=UUID(collector.test_id))
+        if test_run:
```
Comment (bentsi) [inline sct.py:990]:
```suggestion
       if test_run:= ArgusTestRun.get(test_id=UUID(collector.test_id))
        
```

File: sct.py [hunk 2]
```diff
@@ -981,6 +985,21 @@ def collect_logs(test_id=None, logdir=None, backend=None, config_file=None):
...
+            for cluster_type, s3_links in collected_logs.items():
+                for link in s3_links:
+                    if isinstance(link, str):
+                        test_run.run_info.logs.add_log(cluster_type, link)
+                    if isinstance(link, list):
+                        for sub_link in link:
+                            test_run.run_info.logs.add_log(cluster_type, sub_link)
```
Comment (bentsi) [inline sct.py:997]: think of how to put this logic in one method/function

File: sdcm/tester.py [hunk 3]
```diff
@@ -358,6 +363,20 @@ def __init__(self, *args):  # pylint: disable=too-many-statements,too-many-local
...
+    def argus_test_run(self):
+        if not self._argus_test_run and self._use_argus:
+            try:
+                argus_test_run = ArgusTestRun.from_sct_config(test_id=UUID(self.test_id),
+                                                              test_module_path=self.test_config.test_name(),
+                                                              sct_config=self.params)
+                argus_test_run.save()
+                self._argus_test_run = argus_test_run
+            except Exception:  # pylint: disable=broad-except
+                self._use_argus = False
```
Comment (bentsi) [inline sdcm/tester.py:376]: let's return MagicMock() when we encountered exception

File: sdcm/tester.py [hunk 4]
```diff
@@ -358,6 +363,20 @@ def __init__(self, *args):  # pylint: disable=too-many-statements,too-many-local
...
+                self.log.error("Error setting up argus connection", exc_info=True)
+        return self._argus_test_run
+
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sdcm/tester.py]: done

File: sdcm/tester.py [hunk 5]
```diff
@@ -378,6 +397,10 @@ def _init_test_timeout_thread(self) -> threading.Timer:
         self.log.info('Test start time %s, duration is %s and timeout set to %s',
                       start_time, self.test_duration, end_time)
 
+        if self.argus_test_run:
+            self.argus_test_run.run_info.results.status = TestStatus.RUNNING
+            self.argus_test_run.save()
+
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sdcm/tester.py]: Done, I've changed it a bit so the second call takes some time to happen (maybe around 1-5 seconds), but yes, this should be fine

File: sdcm/tester.py [hunk 6]
```diff
@@ -571,6 +594,20 @@ def setUp(self):
         if self.monitors:
             self.monitors.wait_for_init()
 
+        if self.argus_test_run:
+            try:
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sdcm/tester.py]: Done

File: sdcm/tester.py [hunk 7]
```diff
@@ -2252,7 +2290,28 @@ def tearDown(self):
         self.stop_event_device()
         if self.params.get('collect_logs'):
             self.collect_sct_logs()
+
         self.finalize_teardown()
+
+        if self.argus_test_run:
+            try:
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sdcm/tester.py]: Done

File: sdcm/cluster.py [hunk 8]
```diff
@@ -284,6 +287,22 @@ def init(self) -> None:
 
         self.set_keep_alive()
 
+        try:
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sdcm/cluster.py:292]: Done

File: sdcm/cluster.py [hunk 9]
```diff
@@ -1077,6 +1096,15 @@ def destroy(self):
         self.stop_task_threads()
         ContainerManager.destroy_all_containers(self)
         LOGGER.info("%s destroyed", self)
+        try:
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sdcm/cluster.py]: Done

File: sdcm/nemesis.py [hunk 10]
```diff
@@ -2848,6 +2851,20 @@ def wrapper(*args, **kwargs):  # pylint: disable=too-many-statements
 
         with DisruptionEvent(nemesis_name=args[0].get_disrupt_name(),
                              node=args[0].target_node, publish_event=True) as nemesis_event:
+            nemesis_info = None
+            try:
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sdcm/nemesis.py]: Done

File: sdcm/nemesis.py [hunk 11]
```diff
@@ -2895,6 +2913,21 @@ def wrapper(*args, **kwargs):  # pylint: disable=too-many-statements
                 if num_nodes_before != num_nodes_after:
                     args[0].log.error('num nodes before %s and nodes after %s does not match' %
                                       (num_nodes_before, num_nodes_after))
+
+                try:
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sdcm/nemesis.py]: Done

File: sct.py [hunk 12]
```diff
@@ -981,6 +985,21 @@ def collect_logs(test_id=None, logdir=None, backend=None, config_file=None):
 
     click.echo(table.get_string(title="Collected logs by test-id: {}".format(collector.test_id)))
 
+    try:
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sct.py:988]: Done

File: sct.py (hunk 1 above)
Comment (k0machi) [reply sct.py:990]: Done

File: sct.py (hunk 2 above)
Comment (k0machi) [reply sct.py:997]: Done

File: sdcm/tester.py (hunk 3 above)
Comment (k0machi) [reply sdcm/tester.py]: Done

File: sct.py (hunk 2 above)
Comment (k0machi) [reply sct.py]: done

File: sct.py (hunk 1 above)
Comment (k0machi) [reply sct.py]: done

File: sdcm/tester.py [hunk 13]
```diff
@@ -2643,6 +2710,18 @@ def collect_logs(self) -> None:
                     logs_dict[cluster["logname"]] = s3_link
                 else:
                     self.log.warning("There are no logs for %s uploaded", cluster["name"])
+        if self.argus_test_run:
+            try:
```
[the remark this answers is not in the document]
Comment (k0machi) [reply sdcm/tester.py]: done

File: sct.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (k0machi) [reply sct.py:991]: done

**Review by dkropachev — APPROVED** [review]

**Review by dkropachev — APPROVED** [review]

File: sdcm/argus_test_run.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+    @classmethod
+    def get(cls, test_id: UUID = None) -> TestRunWithHeartbeat:
+        if not cls.db_init:
+            cls.init_db()
+
+        if test_id and not cls.TESTRUN_INSTANCE:
+            cls.TESTRUN_INSTANCE = TestRunWithHeartbeat.from_id(test_id)
+
+        if not cls.TESTRUN_INSTANCE:
+            return unittest.mock.MagicMock()
```
Comment (bentsi) [inline sdcm/argus_test_run.py:233]: add warning log message

Comment (k0machi) [reply sdcm/argus_test_run.py:227]: Done

**Review by bentsi — APPROVED** [review]

**Pull request merged**